### ISSUE TITLE
How to define a typed generic base class using Typebox?

### ISSUE THREAD

**Issue by uncvrd** [sinclairzx81/typebox#168]
Hi! First off, thanks for you excellent work on this library. I have a fundamental question that I need a bit help understanding.

In my project I'd like to share a model between my Fastify Backend and another JS application to ensure type safety of a class I'm defining while not having to duplicate code. This currently has me stumped...

**FASTIFY BACKEND**
---

In my Fastify backend, I defined a function to return a Typebox generic model called `GetJobParams` (as instructed by the docs):

```
export const getJobParams = <Q extends TEnumType>(queueNameEnum: Q) => Type.Object({
  id: Type.String(),
  queue: Type.Enum(queueNameEnum)
})
```

I can then extract a typed version of `GetJobParams` by passing my `QueueName` enum to `getJobParams`

```
export enum QueueName {
    MAIN = "main"
}

export const GetJobParams = getJobParams(QueueName);
```

`getJobParams` is exported as a shared method and I can import it in to my other `NodeJS` application

**NODEJS APP**
---

Now in my NodeJS application, I need to create a class which defines a method that "strongly types" the input arguments to only allow values from my `QueueName` enum defined above. I was initially using Typescript interfaces to strongly type this class, but cannot figure out how to use Typebox for this...

I created a generic base class `BullBase` where you can pass in a generic param to define the `Queue` that should be used.

**TS Interface Implementation**
---

```
export class BullBase<TQueueName> extends HTTPDataSource<Context> {
    
    constructor() {}

    async getJob<TJob>({ queue, id }: Bull.GetJobParams<TQueueName>) {
        //....code removed
    }
}
```

You can then use this Base class to create a Typed parent class:

```
export class WatermarkAPI extends BullBase<Bull.Watermark.QueueName> {

    constructor() {
        super();
    }
}
```

**WHAT I TRIED**
---

Within my Base class, I tried to use my `getJobParams` Typebox exported function to define the argument type of `getJob` but was met with TS errors:

```
    async getJob<TJob>({ queue, id }: Static<typeof getJobParams(TQueueName)>) {
        //....code removed
    }
```

<img width="1556" alt="image" src="https://user-images.githubusercontent.com/25097099/162555634-55896ae3-f451-4021-8834-60d675751835.png">

The first one is because I'm trying to pass in a Generic type to the `getJobParams` argument, which is expecting an enum type instead. The second has to do because I'm defining the `Static<>` type all in one line.

Am I thinking about using this library incorrectly? Should I instead just stick to Typescript interfaces? I was hoping to not have to duplicate code but it seems like this scenario isn't what Typebox was built to do.

Basically my question comes down to, in the scenario where you need to define a generic base class to use for strongly typed parent classes, how / would it be possible to use Typebox to accomplish this?

Thanks for any insight!

**Comment by sinclairzx81** [comment]
@uncvrd Hi.

It should be possible to share TypeBox types between projects, however sharing generic function types may cause you some problems, particularly if the generics are constrained to `enum` types (which are difficult to express generically in TS as is). Note, there is actually more work to be done in TypeBox to enable more advanced generic patterns, at this stage I'm currently waiting on some upstream functionality in TypeScript prior to investigating additional generics support.

https://github.com/microsoft/TypeScript/pull/47607

I guess my recommendation here would be to move all types that need sharing (including the `enum`) into the shared project and remove the generics from the shared project. If you do need to export generics from your shared project, I'd recommend using `TUnion<[...TLiteral]>` constraints over `TEnum` as string literal values are easier to express as TypeBox types than enums currently are.

Typically when developing client server applications with TypeBox, ill have the following setup, usually as a mono repository. 

```bash
/libs/client    - browser - imports shared types
/libs/server   - fastify - imports shared types
/libs/shared  - exports shared typebox types live here
```
I have a auxiliary project [here](https://github.com/sinclairzx81/sidewinder) which goes quite deep on schema type sharing between client and server environments (by way of a shared package). This project uses a future version of TypeBox that is being developed to help identify and address some of these cross project type sharing scenarios. You can view an example project [here](https://github.com/sinclairzx81/sidewinder/tree/master/example/basic)

Hope this helps
S

**Comment by uncvrd** [comment]
@sinclairzx81 Thank you for the detailed response!

I should have mentioned, but I did extract my generic shared models like `getJobParams` to a shared `npm` library. Reason was because I ran in to an issue with how to import local shared packages in a docker compose microservices environment. I couldn't use npm workspaces because some microservices weren't JS services (python etc) and also local module resolution in a monorepo would be different between host and a docker container.

I'll take your recommendation and hold off on Typebox generics for now as it seems that the mentioned Typescript ticket will be added in 4.7! 🤞

**Comment by sinclairzx81** [comment]
@uncvrd Hi, yeah can leave this issue open. Will be reviewing enhanced generics support once instantiation expressions lands in TypeScript (and when I have time to review how to integrate). I personally have some downstream use cases (notably shared generic web service contracts in that sidewinder project), will post updates here when I have news to report.

Best of luck!
S

**Comment by sinclairzx81** [comment]
@uncvrd Apologies for the delay in updates on this issue. For your original code example, it looks like you're trying to infer a static type for the generic class argument of `TQueueName`. You can actually do this in TypeBox today, but you need to constrain `TQueueName` to type `TSchema`. 

### Example

I've setup a quick example of how something like this can be achieved with TypeBox, and included a new runtime type checking feature TypeBox now supports. The example below creates a new class `Channel<T>` where `T` is to be expressed by way of TypeBox type. The code then uses static inference to obtain the static type for T, as well as implements the new `TypeCompiler` for type checking. The logic to `read` and `write` has been omitted in the example.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYGELmAGwygBoV1sALDAYwGs4BfOAMylzgHIABAZ2ADtq+AIbAoAehjkARhAAe46rjAEiHAFChIsRGUykAyjGExg1UsgPUqIYYxZsQnXgKGiJUzLLkb148XAAtMEhoWHhEZFR0YF+AVgUwvz8GPgAPMgAfHFBMXn5BWHq6hhy2vBuPDxwCUkp6chwpTAY-AAm1ZbWGLaZiOpwcGBQwABuJhhwUBjCbRD8+Khw3XQAXHqUNLQZ2QPL8zwwUACu1DDQABTDYxNTM3MLSzzdtuvIAJT9g4MwFMA8ADoVvQALwbHB4QhQAEQlSEC6-f4A542YTvPYMYqDMDHaT4MxwYQ8VCCOAAKjgAG0DKgQLJ8ACiSTqABJFpQEzQAC6F0+CD2gwA7n94Udjhg+QLvkp+Ic4ON8OK4GDhILRPBEYDprNeVLBsBmBcFUqQaa4Pxjvh8J9pjBjlB+HB-HAAKIAeQAYnq4KhgKk2vLhIqMFLMYNMXscXiCUzSYKRi0jUHxesjCYzDtJd84AaLgBCTVAqh0GHF2hJ4PvT6-NiC11QNhQC4cePARPvdbGyb-HP8BXANocdHZ8QU85zdZaQggVrwVvtuD4CAAcwJVGm5PEGKxQ1x+OohOJpO1bV56wACo5-hg0mnTNQdnAAD7my34Pr87MyuVd9bHfi0PwECCo6YIIEwo5wOOEDrCeDjsDwED2tQkzrpMZJbtmuYFn8gLAqWWwVuKVZQRQtb1o2zYnmegbBjm1QCP2g7Dt8tr2o6XbbhGzqFLxfGRDkACqPDCMuIY8fxkn8cU6jfhUiTJKkyrmhgda1Ip+AXGgmAAm60gAFY0DAFyfnAchvOQAIAHLHHSRC8sQeyoBZOk2XZTbvI5gwAF4uRg1m2dI9nogwVYyaq6rLAp9QAvOGAmWZ6wAIykM5cAAEykL5cAAMyMOiskHPAKRyPAKpqm2UV1KkALUQVzDQISFXwBccm6HIqVZfYEDMFVGl8k6ASxtQgRtkQnJQMUEZAA)

```typescript
import { TypeCompiler, TypeCheck } from '@sinclair/typebox/compiler'
import { Type, Static, TSchema } from '@sinclair/typebox'

// --------------------------------------------------------
// Channel<T>
// --------------------------------------------------------

export class Channel<T extends TSchema> {
  private readonly check: TypeCheck<T>

  constructor(private readonly schema: T) {
    this.check = TypeCompiler.Compile(this.schema)
  }

  public async * [Symbol.asyncIterator]() {
    while(true) {
      const value = await this.read()
      if(value === null) return // EOF
      yield value
    }
  }

  public async write(value: Static<T>) {
    if(!this.check.Check(value)) throw Error('write(): value is invalid')
    /* todo: implement write() logic here */
  }

  public async read(): Promise<Static<T> | null> {
    const value: unknown = {} /* todo: read from source here */
    if(!this.check.Check(value)) throw Error('read(): value is invalid')
    return value
  }
}

// --------------------------------------------------------
// Usage
// --------------------------------------------------------


const channel = new Channel(Type.Object({
  x: Type.Number(),
  y: Type.Number(),
  z: Type.Number()
}))


await channel.write({ x: 1, y: 2, z: 3 })

const next = await channel.read()

for await (const { x, y, z } of channel) { // async-iterator

}
```

You should be able reshape this example for your usecase. Note there is still an ongoing review of leveraging instantiation expressions in TypeBox, but the above example should help resolve your particular usecase.

Hope this helps :) Will close of this issue for now.

Many Thanks
S
